great_expectations/datasource/fluent/snowflake_datasource.py

```diff
@@ -1,10 +1,21 @@
-from typing import Literal, Union
+from __future__ import annotations
 
+from typing import TYPE_CHECKING, Literal, Optional, Union
+
+import pydantic
 from pydantic import AnyUrl
 
+from great_expectations.compatibility.snowflake import URL
+from great_expectations.compatibility.sqlalchemy import sqlalchemy as sa
 from great_expectations.core._docs_decorators import public_api
 from great_expectations.datasource.fluent.config_str import ConfigStr
-from great_expectations.datasource.fluent.sql_datasource import SQLDatasource
+from great_expectations.datasource.fluent.sql_datasource import (
+    SQLDatasource,
+    SQLDatasourceError,
+)
+
+if TYPE_CHECKING:
+    from great_expectations.compatibility import sqlalchemy
 
 
 class SnowflakeDsn(AnyUrl):
```

Comment:
I think we should ensures that the user has either provided a `connection_string` or the combination of fields to get a working connection.

A pydantic "Root Validator" is probably the best way to do this.
https://docs.pydantic.dev/latest/usage/validators/#root-validators